### PR TITLE
root: configuration reloading

### PR DESCRIPTION
**Pull request by rissson**
EDIT: Configuration watching is now implemented. This PR is best reviewed commit by commit, I'll update this description later.

# Details

This is the first step to handle configuration reloading. With those changes, it is already possible to do so, by sending a SIGUSR2 signal to the Go server process. The next step would be to watch for changes to configuration files and call the Restart function of the GoUnicorn instance.

SIGHUP is catched by the go server and forwarded as-is to gunicorn, which causes it to restart its workers. However, that does not trigger a reload of the Django settings, probably because they are already loaded in the master, before creating any of the worker instances.

SIGUSR2, however, can be used to spawn a new gunicorn master process, but handling it is a bit trickier. Please refer to Gunicorn's documentation[0] for details, especially the "Upgrading to a new binary on the fly" section.

As we are now effectively killing the gunicorn processed launched by the server, we need to handle some sort of check to make sure it is still running. That's done by using the already existing healthchecks, making them useful not only for the application start, but also for its lifetime. If a check is failed too many times in a given time period, the gunicorn processed is killed (if necessary) and then restarted.

[0] https://docs.gunicorn.org/en/20.1.0/signals.html

Other relevant links and documentation:

Python library handling the processing swaping upon a SIGUSR2: https://github.com/flupke/rainbow-saddle/

Golang cannot easily check if a process exists on Unix systems: https://github.com/golang/go/issues/34396

### New Features

- Support file configuration reloading by sending a SIGUSR2 signal to the `server` process.
- The golang server now automatically restarts (gracefully) the underlying gunicorn process if it fails healthchecks.

### Breaking Changes

Not that I'm aware of.

## Checklist

-   [ ] Local tests pass (`ak test authentik/`)
-   [x] The code has been formatted (`make lint-fix`)

If applicable

-   [ ] The documentation has been updated
-   [ ] The documentation has been formatted (`make website`)

### CODE REVIEW

**Comment by rissson** [comment]
I've started working on automatic configuration reloading, but currently the configuration discovery doesn't match between the go and python code, so I'm updating that first.

File: internal/gounicorn/gounicorn.go [hunk 1]
```diff
@@ -96,8 +106,77 @@ func (g *GoUnicorn) healthcheck() {
...
+			break
+		}
+		g.log.Debugf("waiting for new gunicorn pidfile to appear at %s", newPidFile)
+	}
+	if err != nil {
+		g.log.WithError(err).Warning("failed to find the new gunicorn process, aborting")
+		return
+	}
+
+	newPidB, err := ioutil.ReadFile(newPidFile)
```
Comment (rissson) [inline internal/gounicorn/gounicorn.go:145]: If anyone that knows go, knows how to do this without writing 1k lines, I'd be happy to know it too!

File: internal/gounicorn/gounicorn.go [hunk 2]
```diff
@@ -27,6 +34,7 @@ func New() *GoUnicorn {
 	logger := log.WithField("logger", "authentik.router.unicorn")
 	g := &GoUnicorn{
 		log:             logger,
+		pidFile:         nil,
 		started:         false,
 		killed:          false,
```
Comment (rissson) [inline internal/gounicorn/gounicorn.go:39]: Would an enum be better here?

File: internal/gounicorn/gounicorn.go (hunk 1 above)
Comment (rissson) [reply internal/gounicorn/gounicorn.go:145]: `os.ReadFile`

**Review by BeryJu — COMMENTED** [review]
- Configuration changes are good to keep all the defaults in one place
- Gunicorn health checking and reloading is great
- Configuration watching isn't really needed anymore since we can do that natively now
- In a future PR maybe generalise the process manager to be re-usable for Guacamole
  - We also don't really want to have our own "init-system"
  - We could consider using supervisord 
  - Also https://github.com/ochinchina/supervisord exists which we could maybe use as a library

**Pull request closed without merging**